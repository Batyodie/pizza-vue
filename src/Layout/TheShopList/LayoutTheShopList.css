@import "../../style/Base/Extend/BaseExtend.css";

.TheShopList {
  @extend .ContentWrapper;
  max-width: 820px;
}
.Header {
  display: flex;
  width: 100%;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 30px;
}
.Content {
  width: 100%;
  margin: 1px;
}
.Inner {
  display: flex;
  align-items: center;
}
.Title {
  @extend .BaseTitle;
  margin: 15px;
}

.HeaderBtn {
  margin-top: 5px;
  background-color: transparent;
  border: none;
  height: 23px;
  padding: 0;
  & span {
    order: 2;
    color: #b6b6b6;
    font-style: normal;
    font-weight: normal;
    font-size: 16px;
    line-height: 19px;
  }
}

/* footer */
.Footer {
  width: 100%;
}
.FooterStats {
  margin-top: 40px;
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
}
.FooterAllPizas {
  @extend .BaseTitle;
  font-weight: normal;
  font-size: 22px;
  line-height: 27px;
  & span {
    font-weight: bold;
  }
}
.FooterAllSum {
  @extend .FooterAllPizas;
  & span {
    font-weight: bold;
    color: var(--accent-color);
  }
}

.FooterBtns {
  display: flex;
  margin-top: 40px;
  justify-content: space-between;
  flex-wrap: wrap;
}
.FooterBtnBack {
  height: 55px;
  max-height: 55px;
  --btn-color: #cacaca;
  border: 1px solid var(--btn-color);
  background-color: transparent;
  & span {
    order: 2;
    color: var(--btn-color);
    font-weight: normal;
    font-size: 16px;
    line-height: 19px;
    text-align: center;
  }
}
.FooterBtnBackIcon {
  order: 1;
}
.FooterBtnPay {
  height: 55px;
  max-height: 55px;
  margin: 0;
  background-color: var(--accent-color);
  border: 1px solid var(--accent-color);
}

/* empty */
.Empty {
  margin-left: auto;
  margin-right: auto;
  max-width: 550px;
  width: 100%;
  padding: 5px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.EmptyTitle {
  @extend .BaseTitle;
  text-align: center;
}

.EmptyCopy {
  @extend .BaseTitle;
  margin-top: 10px;
  font-weight: normal;
  font-size: 18px;
  line-height: 145.4%;
  text-align: center;
  letter-spacing: 0.01em;
  color: #777777;
}
.EmptyLogo {
  margin-top: 50px;
  max-width: 300px;
  width: 100%;
  max-height: 255px;
  height: 100%;
}
.EmptyBtn {
  margin-top: 60px;
  border: none;
}
@media (max-width: 500px) {
  .FooterBtns {
    justify-content: center;
    align-items: center;
  }
  .FooterBtnBack,
  .FooterBtnPay {
    margin: 5px;
  }
}

@media (max-width: 425px) {
  .TheShopList {
    padding: 20px 30px;
  }
  .FooterBtns {
    justify-content: center;
    align-items: center;
  }
}

@media (max-width: 375px) {
  .TheShopList {
    padding: 10px 10px;
  }
}
