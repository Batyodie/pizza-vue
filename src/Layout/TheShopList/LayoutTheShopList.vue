<!-- Layout the shop list component -->
<script>
import { Button, ShopCart } from "@/components";
import { mapGetters, mapActions } from "vuex";

import { CartIcon, TrashIcon, arrow, EmptyCart } from "@/assets";
import { TheShopListStyle } from "@/Layout/style";
import LayoutTheShopListEmpty from "../TheShopList/LayoutTheShopListEmpty.vue";
export default {
  name: "LayoutTheShopList",
  components: {
    Button,
    ShopCart,
    LayoutTheShopListEmpty
  },

  data() {
    return {
      CartIcon: CartIcon,
      TrashIcon: TrashIcon,
      arrow: arrow,
      EmptyCart: EmptyCart
    };
  },
  computed: {
    TheShopListStyle() {
      return TheShopListStyle;
    },

    ...mapGetters({
      cart: "getCart",
      totalPrice: "getTotalPrice",
      itemsCount: "getItemsCount",
      pizzaItem: "getPizzaItem",
      getBasketFlag: "getBasketFlag"
    })
  },

  methods: {
    ...mapActions({
      removeItems: "removePizzaItems"
    }),
    getClearPizzas() {
      this.$modal.show("dialog", {
        title: "Очистить корзину пицц",
        text: "Вы действительно хотите отчистить корзину",
        buttons: [
          {
            title: "Нет",
            handler: () => {
              this.$modal.hide("dialog");
            }
          },
          {
            title: "Да",
            handler: () => {
              this.removeItems();
            }
          }
        ]
      });
    },

    handlerPayOut() {
      this.$modal.show("dialog", {
        title: "Спасибо за покупку!",
        text: `Вы приобрели ${this.itemsCount} пицц, по цене ${this.totalPrice} рублей.
        \n Спасибо за покупку, приходите еще!`,
        buttons: [
          {
            title: "Закрыть",
            handler: () => {
              this.$modal.hide("dialog");
              console.log(this.cart);
              this.removeItems();
            }
          }
        ]
      });
    }
  }
};
</script>

<template>
  <div :class="TheShopListStyle.TheShopList">
    <template v-if="getBasketFlag">
      <v-dialog />
      <header :class="TheShopListStyle.Header">
        <div :class="TheShopListStyle.Inner">
          <img
            :src="CartIcon"
            :class="TheShopListStyle.CartICON"
            alt="Trash icon"
          />
          <h2 :class="TheShopListStyle.Title">Корзина</h2>
        </div>
        <Button
          @click.native="getClearPizzas"
          :class="TheShopListStyle.HeaderBtn"
        >
          <template slot="Icon">
            <img style="order: 1;" :src="TrashIcon" alt="Trash icon" />
          </template>
          <template slot="ButtonText">
            Очистить корзину
          </template>
        </Button>
      </header>
      <div :class="TheShopListStyle.Content">
        <ShopCart
          v-for="(shopCart, index) in pizzaItem"
          :shopCart="shopCart.items"
          :key="shopCart.id"
          :index="index"
          :groupCartItems="pizzaItem[index].items"
          :groupCartItemsPrice="pizzaItem[index].totalPrice"
        />
      </div>
      <footer :class="TheShopListStyle.Footer">
        <div :class="TheShopListStyle.FooterStats">
          <div>
            <p :class="TheShopListStyle.FooterAllPizas">
              Всего пицц: <span>{{ itemsCount }}</span>
            </p>
          </div>
          <div>
            <p :class="TheShopListStyle.FooterAllSum">
              Сумма заказа: <span>{{ totalPrice }}</span>
            </p>
          </div>
        </div>
        <div :class="TheShopListStyle.FooterBtns">
          <router-link to="/" style="text-decoration: none">
            <Button :class="TheShopListStyle.FooterBtnBack">
              <template slot="ButtonText">
                Вернуться назад
              </template>
              <template slot="Icon">
                <img
                  :class="TheShopListStyle.FooterBtnBackIcon"
                  :src="arrow"
                  alt="arrow left icon"
                />
              </template> </Button
          ></router-link>
          <Button
            @click.native="handlerPayOut"
            :class="TheShopListStyle.FooterBtnPay"
          >
            <template slot="ButtonText">
              Оплатить сейчас
            </template>
          </Button>
        </div>
        <v-dialog name="example" />
      </footer>
    </template>
    <template v-else>
      <LayoutTheShopListEmpty />
    </template>
  </div>
</template>
