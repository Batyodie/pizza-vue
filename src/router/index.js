import Vue from "vue";
import VueRouter from "vue-router";
import Home from "../views/Market";
import Basket from "../views/Basket/Basket";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: "Home",
    component: Home,
  },
  {
    path: "/basket",
    name: "Basket",
    component: Basket,
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  hash: false,
  routes,
});

export default router;
