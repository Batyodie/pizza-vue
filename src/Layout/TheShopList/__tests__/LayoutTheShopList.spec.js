import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import VueRouter from "vue-router";
import router from "@/router";
import Vuex from "vuex";
import VModal from "vue-js-modal";

import LayoutTheShopList from "@/Layout/TheShopList/LayoutTheShopList";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(
  VModal,
  { dialog: true },
  { componentName: "v-dialog-pay", dialog: true }
);
localVue.use(VueRouter);

describe("Unit test from LayoutTheShopList components", () => {
  // fake data
  const pizzaItems = [
    {
      items: [
        {
          id: 7,
          name: "Маргарита",
          imageUrl:
            "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg",
          price: 450,
          size: 26,
          type: 0,
        },
        {
          id: 7,
          name: "Маргарита",
          imageUrl:
            "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg",
          price: 450,
          size: 26,
          type: 0,
        },
      ],
      totalPrice: 900,
      activeBtn: true,
      tags: { type: 0, size: 26 },
    },
    {
      items: [
        {
          id: 8,
          name: "Четыре сезона",
          imageUrl:
            "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/ec29465e-606b-4a04-a03e-da3940d37e0e.jpg",
          price: 395,
          size: 26,
          type: 0,
        },
        {
          id: 8,
          name: "Четыре сезона",
          imageUrl:
            "https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/ec29465e-606b-4a04-a03e-da3940d37e0e.jpg",
          price: 395,
          size: 26,
          type: 0,
        },
      ],
      totalPrice: 790,
      activeBtn: true,
      tags: { type: 0, size: 26 },
    },
  ];
  const totalPrice = 1690;
  const pizzaItemsCount = pizzaItems.length;

  let store;
  let actions;
  let getters;

  beforeEach(() => {
    actions = {
      removePizzaItems: jest.fn(),
      removePizzaItem: jest.fn(),
      plusCartItem: jest.fn(),
      minusCartItem: jest.fn(),
    };
    getters = {
      getCart: () => pizzaItems,
      getTotalPrice: () => totalPrice,
      getItemsCount: () => pizzaItemsCount,
      getPizzaItem: () => pizzaItems,
      getBasketFlag: () => (totalPrice === 0 ? false : true),
    };

    store = new Vuex.Store({
      actions,
      getters,
    });
  });

  // create wrapper factory
  const wrapperFactory = (mouning = shallowMount, { propsData } = {}) => {
    return mouning(LayoutTheShopList, {
      localVue,
      store,
      propsData: {
        ...propsData,
      },
      router,
    });
  };

  it("should initialize correctly layoutTheGrid component", () => {
    const wrapper = wrapperFactory(mount);

    // correctly Vue instance
    expect(wrapper.exists()).toBeTruthy();
    // We expect the wrapper to be layout the grid component
    expect(wrapper.is(LayoutTheShopList)).toBeTruthy();
  });

  it("should correctly first render  layoutTheGrid component", () => {
    const wrapper = wrapperFactory(mount);
    let index = 0;
    const shopCards = wrapper.findAll(".ShopCart");
    const renderTotalPrice = `Сумма заказа: ${totalPrice}`;
    const renderTotalCount = `Всего пицц: ${pizzaItemsCount}`;

    shopCards.wrappers.forEach((e, i) => {
      const dataURL = pizzaItems[index].items[index];
      const renderCardTitle = `${dataURL.name}`;
      const renderCardSubtitle = `${
        dataURL.type === 0 ? "тонкое" : "традиционное"
      } тесто, ${dataURL.size} см.`;

      const renderCardImg = dataURL.imageUrl;
      const renderCardPrice = `${pizzaItems[index].totalPrice} ₽`;

      if (index === i) {
        const elem = shopCards.at(index);

        expect(elem.find(".Title").text()).toEqual(renderCardTitle);
        expect(elem.find(".SubTitle").text()).toEqual(renderCardSubtitle);
        expect(
          elem.find(".ShopCart > picture > img").attributes("src")
        ).toEqual(renderCardImg);
        expect(elem.find(".TotalPrice").text()).toEqual(renderCardPrice);
        expect(elem.find(".Counter > span").text()).toBe(
          String(pizzaItems[index].items.length)
        );
      }
    });
    expect(wrapper.find(".FooterAllPizas").text()).toEqual(renderTotalCount);
    expect(wrapper.find(".FooterAllSum").text()).toEqual(renderTotalPrice);
  });

  it("should match the snapshot layoutTheGrid component", () => {
    const wrapper = wrapperFactory(mount);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("should correctly call action removePizzaItem in layoutTheGrid component", async () => {
    const wrapper = wrapperFactory(mount);
    const removeItemBtn = wrapper.findAll(".ResetBtn");
    await removeItemBtn.at(0).trigger("click");
    await wrapper
      .findAll(".vue-dialog-buttons > button")
      .at(1)
      .trigger("click");
    expect(actions.removePizzaItem).toHaveBeenCalled();
  });
  it("should correctly call action removePizzaItems in layoutTheGrid component", async () => {
    const wrapper = wrapperFactory(mount);
    const removeItemsBtn = wrapper.find(".HeaderBtn");
    await removeItemsBtn.trigger("click");
    await wrapper
      .findAll(".vue-dialog-buttons > button")
      .at(1)
      .trigger("click");
    expect(actions.removePizzaItems).toHaveBeenCalled();
  });
  it("should correctly call action plusCartItem in layoutTheGrid component", async () => {
    const wrapper = wrapperFactory(mount);
    const plusBtn = wrapper.findAll(".CounterBtn").at(1);
    await plusBtn.trigger("click");

    expect(actions.plusCartItem).toHaveBeenCalled();
  });
  it("should correctly call action minusCartItem in layoutTheGrid component", async () => {
    const wrapper = wrapperFactory(mount);
    const minusBtn = wrapper.findAll(".CounterBtn").at(0);
    await minusBtn.trigger("click");

    expect(actions.minusCartItem).toHaveBeenCalled();
  });
  it("should correctly call action pay pizzas in layoutTheGrid component", async () => {
    const wrapper = wrapperFactory(mount);
    const btnPay = wrapper.find(".FooterBtnPay");
    await btnPay.trigger("click");
    await wrapper
      .findAll(".vue-dialog-buttons > button")
      .at(0)
      .trigger("click");
    expect(actions.removePizzaItems).toHaveBeenCalled();
  });
});
