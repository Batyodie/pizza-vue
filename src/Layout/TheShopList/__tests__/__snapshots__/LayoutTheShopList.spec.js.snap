// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test from LayoutTheShopList components should match the snapshot layoutTheGrid component 1`] = `
<div class="TheShopList">
  <!---->
  <header class="Header">
    <div class="Inner"><img src="" alt="Trash icon" class="CartICON">
      <h2 class="Title">Корзина</h2>
    </div> <button type="button" class="Btn HeaderBtn"><span class="Body">
          Очистить корзину
        </span> <img src="" alt="Trash icon" style="order: 1;"></button>
  </header>
  <div class="Content">
    <div class="ShopCart">
      <picture>
        <source srcset="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg" width="80" height="80" type="image/svg"> <img src="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/d48003cd-902c-420d-9f28-92d9dc5f73b4.jpg" width="80" height="80" alt="ShopCart pizza image">
      </picture>
      <div class="Typh">
        <h3 class="Title">Маргарита</h3>
        <p class="SubTitle">
          тонкое тесто, 26 см.
        </p>
      </div>
      <div class="Counter"><button type="button" class="Btn CounterBtn"><span class="Body">
        -
      </span> </button> <span style="margin: 0px 12px 0px 12px;">
      2
    </span> <button type="button" class="Btn CounterBtn"><span class="Body">
        +
      </span> </button></div>
      <div class="TotalContainer">
        <div class="TotalPrice"><span> 900 ₽ </span></div>
      </div>
      <div class="Reset"><button type="button" class="Btn ResetBtn"><span class="Body">
        x
      </span> </button>
        <!---->
      </div>
    </div>
    <div class="ShopCart">
      <picture>
        <source srcset="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/ec29465e-606b-4a04-a03e-da3940d37e0e.jpg" width="80" height="80" type="image/svg"> <img src="https://dodopizza.azureedge.net/static/Img/Products/Pizza/ru-RU/ec29465e-606b-4a04-a03e-da3940d37e0e.jpg" width="80" height="80" alt="ShopCart pizza image">
      </picture>
      <div class="Typh">
        <h3 class="Title">Четыре сезона</h3>
        <p class="SubTitle">
          тонкое тесто, 26 см.
        </p>
      </div>
      <div class="Counter"><button type="button" class="Btn CounterBtn"><span class="Body">
        -
      </span> </button> <span style="margin: 0px 12px 0px 12px;">
      2
    </span> <button type="button" class="Btn CounterBtn"><span class="Body">
        +
      </span> </button></div>
      <div class="TotalContainer">
        <div class="TotalPrice"><span> 790 ₽ </span></div>
      </div>
      <div class="Reset"><button type="button" class="Btn ResetBtn"><span class="Body">
        x
      </span> </button>
        <!---->
      </div>
    </div>
  </div>
  <footer class="Footer">
    <div class="FooterStats">
      <div>
        <p class="FooterAllPizas">
          Всего пицц: <span>2</span></p>
      </div>
      <div>
        <p class="FooterAllSum">
          Сумма заказа: <span>1690</span></p>
      </div>
    </div>
    <div class="FooterBtns"><a href="/" aria-current="page" class="router-link-exact-active router-link-active" style="text-decoration: none;"><button type="button" class="Btn FooterBtnBack"><span class="Body">
              Вернуться назад
            </span> <img src="" alt="arrow left icon" class="FooterBtnBackIcon"></button></a> <button type="button" class="Btn FooterBtnPay"><span class="Body">
            Оплатить сейчас
          </span> </button></div>
    <!---->
  </footer>
</div>
`;
